File: frontend/packages/api/src/grpc/validateQuery.ts
import type { ValidateResult } from '@sql-client/types'
import { getGrpcClient } from '@api/grpc/client'
import { mapGrpcError } from '@api/grpc/errors'

export async function validateQuery(baseUrl: string, sql: string): Promise<ValidateResult> {
  try {
    const client = getGrpcClient(baseUrl)
    const call = client.validateQuery({ sql })
    const response = await call.response
    return {
      valid: response.valid,
      message: response.message,
    }
  } catch (error) {
    throw mapGrpcError(error)
  }
}

File: backend/sql-client/internal/grpc/validate_query.go
package grpc

import (
	"context"

	sqlclientv1 "github.com/sql-client/sql-api/gen/sqlclient/v1"
	"github.com/sql-client/sql-api/internal/sqlvalidator"
)

func (h *Handler) ValidateQuery(_ context.Context, req *sqlclientv1.ValidateQueryRequest) (*sqlclientv1.ValidateQueryResponse, error) {
	if err := sqlvalidator.Validate(req.GetSql()); err != nil {
		return &sqlclientv1.ValidateQueryResponse{
			Valid:   false,
			Message: err.Error(),
		}, nil
	}

	return &sqlclientv1.ValidateQueryResponse{Valid: true}, nil
}